rfcs/0018-glides-trainsheet-events.md

```diff
@@ -140,69 +105,125 @@ In order to reduce event duplication, a Trip Key is used to identify both added
-- `consist` (array of Delta Car, optional): the cars assigned to perform this trip. If a car is `"cleared"`, then no car is currently assigned to that position in the consist. If a car is `"unmodified"` then the car is the value from the most-recent [Trip updated](#Trip-updated) event. The front car is listed first.
-- `scheduledOperators`: (array of Operator, optional): if present, the operators who were scheduled to operate this trip.
-- `operators` (array of Delta Operator, optional): the list of operators assigned to perform this trip. If an operator is `"cleared"`, then no operator is assigned to the car in that position in the consist. If the operator is `"unmodified"` then the operator is the value from the schedule or the most-recent [Trip updated](#Trip-updated) event. The operator of the front car is listed first.
-- `dropped`: (Dropped Reason or `false`, optional): if `false`, the trip is not dropped (and restored if previously dropped). If a [Dropped Reason](#dropped-reason), the reason the trip was dropped.
+- `cars` (array of [Car](#Car), optional): array of length 1 or 2, containing the car numbers and operators for each car in the train assigned to the trip. If absent, there are no changes to any cars. If present, the length of the array is the length of the train. In a two car train, the front car is listed first.
+- `dropped` (boolean, optional): whether the trip has been dropped. If `false`, the trip is not dropped (and restored if previously dropped). Added trips can be dropped and restored just like scheduled trips.
+- `droppedReason` ([DroppedReason](#DroppedReason) | null, optional): The reason the trip was dropped. SHOULD only exist if the trip has been dropped. SHOULD be `null` if the trip was restored.
+- `scheduled` ([Scheduled](#Scheduled) | null): For trips in Glides' schedule, this represents the scheduled data for the trip. It will always be present and won't change between subsequent updates to the same trip. For added trips, this will always be null.
 
 *Notes*
 - setting a new location or time does not modify the [Trip Key](#trip-key) for a scheduled trip.
-- `consist` and `operators` MUST be the length of the train: length 1 for a 1-car consist, length 2 for a 2-car consist.
+- if a trip is dropped, fields that are then irrelevant (such as `startTime` or `cars`) SHOULD NOT be set.
+- If `cars` is updated from a 1-element list to a 2-element list, any unset fields in the 2nd car should be considered `"cleared"`. This is relevant if a two-car train has a car removed and then re-added. The 2nd car does not retain the data it used to have (and doesn't fall through to the scheduled operator) unless that data is specifically included in the new object.
+
+
+#### TripAdded
+There is a new trip, that does not appear in the Glides' schedule data. Any added trips will appear exactly once a TripAdded object, and any future updates will appear as TripUpdated objects.
```

Comment:
Oh! good point. These events should all be idempotent, and this bit currently is not. I'll reword it.

Reply:
I tried checking that everything is idempotent, but ran into a problem.

Most events are already idempotent:
- trip updates are inherently idempotent, because they're setting fields.
- adding trips just needs the caveat to check the trip key.
- editor changes are idempotent if you allow stopping editing when you hadn't started, which is already required in the spec.

But sign ins can't be idempotent. How would a client distinguish an operator signing in twice a row from an operator signing in once and getting replayed? Do we have to change the semantics from "you can know all past sign ins" to "you can only know an operator's most recent sign in time"?

That's what the event `id` field is for, right? How is that `id` generated? Deduplicating replayed events should happen somewhere between Glides and Kinesis, not by the Client.

Reply:
clarified in https://github.com/mbta/technology-docs/pull/18/commits/7097cdf9193c6a5b5a12094b2c3660e7b45e89f1